src/metabase/query_processor/expand.clj

```diff
@@ -307,7 +307,7 @@
       :next    (recur f  1 unit))
     (let [f (datetime-field f unit)]
       (cond
-        (core/= n  0) (= f (value f (relative-datetime :current)))
+        (core/= n  0) (= f (value f (relative-datetime  0 unit)))
```

Comment:
Basically after the query is expanded, the field that contains 0 (week, month year) is replaced by :current day which is why Druid filters on today instead of this week/month/year.

I tried to test a bit and see if this has an effect on SQL queries but haven't found any. Filters on the sample dataset seem correct.
However im really not familiar with this part of code, so a deep look is required.